core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala

```diff
@@ -99,43 +117,144 @@ class KubernetesClient(
+      // Therefore we must enumerate them and ensure they are unpaused before we attempt to delete them.
+      Future {
+        blocking {
+          kubeRestClient
+            .inNamespace(config.namespace)
+            .pods()
+            .withLabel(key, value)
+            .list()
+            .getItems
+            .asScala
+            .map { pod =>
+              val container = toContainer(pod)
+              container
+                .resume()
+                .recover { case _ => () } // Ignore errors; it is possible the container was not actually suspended.
+                .map(_ => rm(container))
+            }
+        }
+      }.flatMap(futures =>
+        Future
+          .sequence(futures)
+          .map(_ => ()))
+    } else {
+      runCmd(Seq("delete", "--now", "pod", "-l", s"$key=$value"), config.timeouts.rm).map(_ => ())
     }
   }
 
-  def rm(id: ContainerId)(implicit transid: TransactionId): Future[Unit] =
-    runCmd(Seq("delete", "--now", "pod", id.asString), timeouts.rm).map(_ => ())
+  def suspend(container: KubernetesContainer)(implicit transid: TransactionId): Future[Unit] = {
+    if (config.invokerAgent.enabled) {
+      agentCommand("suspend", container)
+        .map { response =>
+          response.discardEntityBytes()
+        }
+    } else {
+      Future.successful({})
+    }
+  }
 
-  def rm(key: String, value: String)(implicit transid: TransactionId): Future[Unit] =
-    runCmd(Seq("delete", "--now", "pod", "-l", s"$key=$value"), timeouts.rm).map(_ => ())
+  def resume(container: KubernetesContainer)(implicit transid: TransactionId): Future[Unit] = {
+    if (config.invokerAgent.enabled) {
+      agentCommand("resume", container)
+        .map { response =>
+          response.discardEntityBytes()
+        }
+    } else {
+      Future.successful({})
+    }
+  }
 
-  def logs(id: ContainerId, sinceTime: Option[Instant], waitForSentinel: Boolean = false)(
+  def logs(container: KubernetesContainer, sinceTime: Option[Instant], waitForSentinel: Boolean = false)(
     implicit transid: TransactionId): Source[TypedLogLine, Any] = {
 
     log.debug(this, "Parsing logs from Kubernetes Graph Stage…")
 
     Source
-      .fromGraph(new KubernetesRestLogSourceStage(id, sinceTime, waitForSentinel))
+      .fromGraph(new KubernetesRestLogSourceStage(container.id, sinceTime, waitForSentinel))
       .log("foobar")
 
   }
 
+  private def toContainer(pod: Pod): KubernetesContainer = {
+    val id = ContainerId(pod.getMetadata.getName)
+    val addr = ContainerAddress(pod.getStatus.getPodIP)
+    val workerIP = pod.getStatus.getHostIP
+    // Extract the native (docker or containerd) containerId for the container
+    // By convention, kubernetes adds a docker:// prefix when using docker as the low-level container engine
+    val nativeContainerId = pod.getStatus.getContainerStatuses.get(0).getContainerID.stripPrefix("docker://")
```

Comment:
Would it make sense to make the invoker-agent take care of the low-level details, and make the logic here agnostic of specific container engine implementation? For example, if explicit docker interaction is replaced with CRI?

Reply:
not a bad idea, but let's defer that to a later PR.  I'd like to get this one merged quickly, so I can submit the logging implementation I have that builds on it. Then can rethink how we pass the nativeContainerId to the invokerAgent.